subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskRequiredServices.java

```diff
@@ -58,10 +54,9 @@ public DefaultTaskRequiredServices(TaskInternal task, TaskMutator taskMutator, P
 
     @Override
     public Set<Provider<? extends BuildService<?>>> getElements() {
-        if (requiredServices == null) {
```

Comment:
Or we will would have a stale collection if we retrieved elements before they were all added